File: src/glyphs/provenance/decorator/prototypes/check_fixtures.py
"""Runs nfprov_html.py against ../fixtures.json. Usage: .venv/bin/python check_fixtures.py"""
import json, os, sys
HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
import nfprov_html as N
fx = json.load(open(os.path.join(HERE, '..', 'fixtures.json'), encoding='utf-8'))
fail = skip = 0
for c in fx['cases']:
    if 'merge_whitespace' in c['options']:
        skip += 1  # the prototype predates this option
        continue
    got = [{'state': s, 'text': t} for s, t in N.runs(c['input'], strip=c['options'].get('strip', False))]
    if got != c['runs']:
        fail += 1
        print('FAIL', c['name'], '\n got', got, '\n exp', c['runs'])
print(f"{len(fx['cases']) - fail - skip}/{len(fx['cases'])} pass, {skip} skipped (merge_whitespace)")
sys.exit(1 if fail else 0)

File: src/glyphs/provenance/decorator/prototypes/check_fixtures.mjs
// Runs the prototype nfprov.js against ../fixtures.json. Usage: node check_fixtures.mjs
// The shipped css/nfprov.js is checked by ../check_fixtures.mjs.
import fs from 'fs'; import path from 'path'; import { fileURLToPath } from 'url';
const here = path.dirname(fileURLToPath(import.meta.url));
const src = fs.readFileSync(path.join(here, 'nfprov.js'), 'utf8').replace('})(window);', '})(globalThis);');
(0, eval)(src);
const fx = JSON.parse(fs.readFileSync(path.join(here, '..', 'fixtures.json'), 'utf8'));
// Cases the prototype is known to fail. It marks every plane-16 code point,
// not only those in mapping.json; css/nfprov.js fixed this.
const KNOWN_FAILURES = new Set(['pua code points outside the mapping pass through unmarked']);
let fail = 0, skip = 0;
for (const c of fx.cases) {
  if ('merge_whitespace' in c.options || KNOWN_FAILURES.has(c.name)) { skip++; continue; } // predates the option / known gap
  const got = nfprov.runs(c.input, c.options);
  const ok = JSON.stringify(got) === JSON.stringify(c.runs);
  if (!ok) { fail++; console.log('FAIL', c.name, '\n got', JSON.stringify(got), '\n exp', JSON.stringify(c.runs)); }
}
console.log(`${fx.cases.length - fail - skip}/${fx.cases.length} pass, ${skip} skipped (merge_whitespace, known gaps)`);
process.exit(fail ? 1 : 0);
